src/pages/academy/xpCalculation/XpCalculation.tsx

```diff
@@ -0,0 +1,249 @@
+import 'ag-grid-community/dist/styles/ag-grid.css';
+import 'ag-grid-community/dist/styles/ag-theme-balham.css';
+
+import { Button, Divider, NonIdealState, Spinner, SpinnerSize } from '@blueprintjs/core';
+import { IconNames } from '@blueprintjs/icons';
+import { GridApi, GridReadyEvent } from 'ag-grid-community';
+import { AgGridReact } from 'ag-grid-react';
+import { useEffect, useState } from 'react';
+import ContentDisplay from 'src/commons/ContentDisplay';
+
+export type StateProps = {
+  allUserXp: string[][] | undefined;
+};
+
+export type DispatchProps = {
+  handleAllUserXpFetch: () => void;
+};
+
+type XpCalculationProps = DispatchProps & StateProps;
+
+type RowData = {
+  name: string;
+  'NUS Net ID': string;
+  'Assessment Xp': string;
+  'Achievement Xp': string;
+};
+
+type PageState = {
+  currPage: number;
+  maxPages: number;
+  rowCountString: string;
+  isBackDisabled: boolean;
+  isForwardDisabled: boolean;
+};
+
+const XpCalculation = ({ allUserXp, handleAllUserXpFetch }: XpCalculationProps) => {
+  const [rowData, setRowData] = useState<RowData[]>([]);
+  const [gridApi, setGridApi] = useState<GridApi>();
+  const [pageState, setPageState] = useState<PageState>({
+    currPage: 1,
+    maxPages: 1,
+    rowCountString: '(none)',
+    isBackDisabled: true,
+    isForwardDisabled: true
+  });
+  // const [filterText, setFilterText] = useState<string>('');
+
+  useEffect(() => {
+    if (allUserXp) {
+      const rowData = allUserXp.map(data => {
+        const [name, nusNetId, assessmentXp, achievementXp] = data;
+        return {
+          name: name ?? '',
+          'NUS Net ID': nusNetId ?? '',
+          'Assessment Xp': assessmentXp ?? '0',
+          'Achievement Xp': achievementXp ?? '0'
+        };
+      });
+      setRowData(rowData);
+    }
+  }, [allUserXp]);
+
+  const columnDefs = [
+    { field: 'name' },
+    { field: 'NUS Net ID' },
+    { field: 'Assessment Xp' },
+    { field: 'Achievement Xp' }
+  ];
+
+  const defaultColDef = {
+    resizable: true,
+    sortable: true
+  };
```

Comment:
These constants don't depend on anything; moving them out of the component (so that a shared copy is used instead of creating a new object each time) may improve performance.

Reply:
Moreover, marking them as read-only using `as const` may also be appropriate.